source/module_rdmft/rdmft_tools.cpp

```diff
@@ -0,0 +1,414 @@
+                    const std::string XC_func_rdmft, 
+                    const double alpha)
+{ 
+    occNum_wfcHwfc.zero_out();
+    occNum_Mul_wfcHwfc(occ_number, wfcHwfc_exx_XC_in, occNum_wfcHwfc, 4, XC_func_rdmft, alpha);
+    occNum_wfcHwfc+=(wfcHwfc_TV_in);
+    occNum_wfcHwfc+=(wfcHwfc_hartree_in);
+    occNum_wfcHwfc+=(wfcHwfc_dft_XC_in);
+
+    // consider W_k for dE/d_occNum
+    for(int ik=0; ik<occ_number.nr; ++ik)
+    {
+        for(int inb=0; inb<occ_number.nc; ++inb) { occNum_wfcHwfc(ik, inb) *= kv.wk[ik];
+}
+    } 
+}
+
+
+// do wk*g(occNum)*wfcHwfc and add for TV, hartree, XC. This function just use once, so it can be replace and delete
+void add_wfcHwfc(const ModuleBase::matrix& wg, 
+                    const ModuleBase::matrix& wk_fun_occNum, 
+                    const ModuleBase::matrix& wfcHwfc_TV_in, 
+                    const ModuleBase::matrix& wfcHwfc_hartree_in,
+                    const ModuleBase::matrix& wfcHwfc_XC_in, 
+                    ModuleBase::matrix& occNum_wfcHwfc, 
+                    const std::string XC_func_rdmft, 
+                    const double alpha)
+{
+    occNum_wfcHwfc.zero_out();
+    occNum_Mul_wfcHwfc(wg, wfcHwfc_TV_in, occNum_wfcHwfc);
+    occNum_Mul_wfcHwfc(wg, wfcHwfc_hartree_in, occNum_wfcHwfc, 1);
+    occNum_Mul_wfcHwfc(wk_fun_occNum, wfcHwfc_XC_in, occNum_wfcHwfc, 1);
+}
+
+
+// give certain occNum_wfcHwfc, get the corresponding energy
+double getEnergy(const ModuleBase::matrix& occNum_wfcHwfc)
+{
+    double energy = 0.0;
+    for(int ir=0; ir<occNum_wfcHwfc.nr; ++ ir)
+    {
+        for(int ic=0; ic<occNum_wfcHwfc.nc; ++ic) { energy += occNum_wfcHwfc(ir, ic);
+}
+    }
+    return energy;
+}
+
+
+// for HF, Muller and power functional, g(eta) = eta, eta^0.5, eta^alpha respectively.
+// when symbol = 0, 1, 2, 3, 4, 5, return eta, 0.5*eta, g(eta), 0.5*g(eta), d_g(eta)/d_eta, 1.0 respectively.
+// Default symbol=0, XC_func_rdmft="HF", alpha=0.656
+double occNum_func(const double eta, const int symbol, const std::string XC_func_rdmft, double alpha)
+{
+    // if( XC_func_rdmft == "hf" || XC_func_rdmft == "default" || XC_func_rdmft == "pbe0" ) alpha = 1.0;
+    // else if( XC_func_rdmft == "muller" ) alpha = 0.5;
+    // else if( XC_func_rdmft == "power" || XC_func_rdmft == "wp22" || XC_func_rdmft == "cwp22" ) ;
+    // else alpha = 1.0;
+    if( XC_func_rdmft == "power" || XC_func_rdmft == "wp22" || XC_func_rdmft == "cwp22" ) { ;
+    } else if( XC_func_rdmft == "muller" ) { alpha = 0.5;
+    } else { alpha = 1.0;
+}
+
+    if( symbol==0 ) { return eta;
+    } else if ( symbol==1 ) { return 0.5*eta;
+    } else if ( symbol==2 ) { return std::pow(eta, alpha);
+    } else if ( symbol==3 ) { return 0.5*std::pow(eta, alpha);
+    } else if ( symbol==4 ) { return alpha*std::pow(eta, alpha-1.0);
+    } else if ( symbol==5 ) { return 1.0;
+    } else 
+    {
+        std::cout << "\n!!!!!!\nThere may be some errors when calling wg_fun()\n!!!!!!\n";
```

Comment:
what is the behavior you really expect here? Do you want to trigger an assertion failure here? if so, you use assert(symbol <= 5) instead.